numba/np/ufunc/deviceufunc.py

```diff
@@ -732,110 +724,164 @@ def _broadcast_scalar_input(self, ary, shape):
+        """
+        Copy `hostary` to the device and return the device array.
+        """
+
+    @abstractmethod
+    def device_array(self, shape, dtype):
+        """
+        Allocate a new uninitialized device array with the given shape and
+        dtype.
+        """
+
     def __init__(self, nin, nout, args, kwargs):
-        if nout > 1:
-            raise ValueError('multiple output is not supported')
-        self.args = args
-        self.kwargs = kwargs
-
-        user_output_is_device = False
-        self.output = self.kwargs.get('out')
-        if self.output is not None:
-            user_output_is_device = self.is_device_array(self.output)
-            if user_output_is_device:
-                self.output = self.as_device_array(self.output)
-        self._is_device_array = [self.is_device_array(a) for a in self.args]
-        self._need_device_conversion = (not any(self._is_device_array) and
-                                        not user_output_is_device)
-
-        # Normalize inputs
-        inputs = []
-        for a, isdev in zip(self.args, self._is_device_array):
-            if isdev:
-                inputs.append(self.as_device_array(a))
+        # Get outputs either from the out kwarg, or generate an initial list of
+        # "placeholder" outputs using None as a sentry value
+        outputs = kwargs.get('out')
+        if outputs is not None and len(args) > nin:
+            raise ValueError("cannot specify 'out' as both positional "
+                             "and keyword arguments")
+        else:
+            outputs = [outputs] * nout
+
+        # Ensure all output device arrays are Numba device arrays - any output
+        # passed in that supports the CUDA Array Interface is converted; others
+        # are left untouched.
+        all_user_outputs_are_host = True
+        self.outputs = []
+        for output in outputs:
+            if self.is_device_array(output):
+                self.outputs.append(self.as_device_array(output))
+                all_user_outputs_are_host = False
             else:
-                inputs.append(np.asarray(a))
-        self.norm_inputs = inputs[:nin]
+                self.outputs.append(output)
+
+        all_host_arrays = not any([self.is_device_array(a) for a in args])
+
+        # - If any of the arguments are device arrays, we leave the output on
+        #   the device.
+        self._copy_result_to_host = (all_host_arrays and
+                                     all_user_outputs_are_host)
+
+        # Normalize inputs - ensure they are either device- or host-side arrays
+        # (as opposed to lists, tuples, etc).
+        def normalize_input(a):
+            if self.is_device_array(a):
+                convert = self.as_device_array
+            else:
+                convert = np.asarray
+
+            return convert(a)
+
+        normalized_inputs = [normalize_input(a) for a in args]
+        self.inputs = normalized_inputs[:nin]
```

Comment:
Is this side-effect normalising the potential outputs specified in `args[-nout:]` ?

Reply:
I *think* the code is confusing due to the names of things here - everything, in and out, should be normalized. Probably `normalize_input()` should be called `normalize_arg()`, and `normalized_inputs` should be called `normalized_args`.

Reply:
I've now renamed these.

Reply:
Thanks, change looks good.